src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -0,0 +1,580 @@
+
+                mantissa = topBits + middleBits + bottomBits;
+
+                uint unusedBottomBlockBitsMask = (1u << (int)(topBlockBits)) - 1;
+                hasZeroTail &= (bottomBlock & unusedBottomBlockBitsMask) == 0;
+            }
+
+            for (uint i = 0; i != bottomBlockIndex; i++)
+            {
+                hasZeroTail &= (value.GetBlock(i) == 0);
+            }
+
+            return AssembleFloatingPointBits(in info, mantissa, exponent, hasZeroTail);
+        }
+
+        // get 32-bit integer from at most 9 digits
+        private static uint DigitsToInt(char* p, int count)
+        {
+            Debug.Assert((1 <= count) && (count <= 9));
+
+            char* end = (p + count);
+            uint res = (uint)(p[0] - '0');
+
+            for (p++; p<end; p++)
+            {
+                res = (10 * res) + p[0] - '0';
+            }
+
+            return res;
+        }
+
+        private static ulong NumberToFloatingPointBitsRoslyn(ref NumberBuffer number, in FloatingPointInfo info)
+        {
+            char* src = number.GetDigitsPointer();
+            uint totalDigits = (uint)(number.Precision);
+
+            Debug.Assert(src[0] != '0');
+
+            if (totalDigits == 0)
+            {
+                return info.ZeroBits;
+            }
+
+            // To generate an N bit mantissa we require N + 1 bits of precision.  The  
+            // extra bit is used to correctly round the mantissa (if there are fewer bits  
+            // than this available, then that's totally okay; in that case we use what we  
+            // have and we don't need to round).
+            uint requiredBitsOfPrecision = (uint)(info.NormalMantissaBits + 1);
+
+            // The input is of the form 0.Mantissa x 10^Exponent, where 'Mantissa' are  
+            // the decimal digits of the mantissa and 'Exponent' is the decimal exponent.  
+            // We decompose the mantissa into two parts: an integer part and a fractional  
+            // part.  If the exponent is positive, then the integer part consists of the  
+            // first 'exponent' digits, or all present digits if there are fewer digits.  
+            // If the exponent is zero or negative, then the integer part is empty.  In  
+            // either case, the remaining digits form the fractional part of the mantissa.  
+            uint positiveExponent = (uint)(Math.Max(0, number.Scale));
+            uint integerDigitsPresent = Math.Min(positiveExponent, totalDigits);
+            uint integerDigitsMissing = positiveExponent - integerDigitsPresent;
+            uint integerFirstIndex = 0;
+            uint integerLastIndex = integerDigitsPresent;
+
+            uint fractionalFirstIndex = integerLastIndex;
+            uint fractionalLastIndex = totalDigits;
+            uint fractionalDigitsPresent = fractionalLastIndex - fractionalFirstIndex;
+
+            // When the number of significant digits is less than or equal to 15 and the
```

Comment:
For `System.Single`, this covers a very good range of possible inputs. You only need at most 9 digits to guarantee round-tripping and the worst case is `1.4 * 10^-45`. This means that, for an expected range of inputs, we will likely always be faster than previous, and there will be some very small/large numbers that are slower.

For `System.Double`, this likewise covers a good range of possible inputs. However, you need at most 17-digits to guarantee round-tripping so there are likely some round-trippable inputs that will always be slower (but that will be correct now). There will also be a much broader range of values that will be slower, since the worst case is `4.9 * 10^-324`, but these fall outside the range of what I would expect to be "normal" inputs (IIRC, approx half of all unique values fall between `-1.0 and 1.0`, due to the representation, and for those, I think we can assume the magnitude of most values will be larger than `1e-22` and less than `1e22`).
* There may be some other tricks we can do to speed up this more for things like the 17 digit case. For example: the "golden standard" by David M. Gay, uses a software implementation of an "extended-precision" float with 96-bits to cover some of these inputs (we would need to do all the appropriate license checks before using it, however, other languages like Java use a variation of it so it probably won't be too much of a concern).

Reply:
If I did my math right, we can have a couple other fast paths to cover the following additional ranges. Doing so would provide code similar to the previous implementation, which was providing a software implementation for 80-bit extended-precision multiply; however, we would also need to provide a software-based division implementation, since negative exponents are not exactly representable.

* System.Single (8-bit exponent, 24-bit mantissa)
  * 7 Digits, "fast exponent" up to 10
* System.Double (11-bit exponent, 53-bit mantissa)
  * 15 Digits, "fast exponent" up to 22
* Extended-Precision Double (15-bit exponent, 64-bit mantissa)
  * 19 Digits, "fast exponent" up to 27
* Quad-Precision (15-bit exponent, 113-bit mantissa)
  * 34 Digits, "fast exponent" up to 38

Supporting a basic software implementation for "extended-precision double" is more trivial since it basically just relies on things we have existing hardware support for (such as `64*64=128`). This would also cover most doubles that are of "round-trippable" length (17 digits).

Supporting a basic software implementation for "quad-precision" is more complex, but may still be worthwhile, since it would effectively cover the entire range of `System.Single` "normal" inputs (for `System.Single`, only subnormal inputs have an exponent that is greater than 38).

Reply:
It now appropriately uses `System.Single` or `System.Double` arithmetic, where applicable.

It also uses `Extended-Precision` arithmetic for the multiplication case. For the division case (which is arguably more common), the implementation isn't as trivial. Berkeley has an existing software-implementation (available under the BSD 3 Clause) for most of the floating-point functions (for binary16, binary32, binary64, binary80, and binary128), which would be nice if we could use here (rather than needing to port our own): https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/COPYING.txt